{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-11-13T14:24:39.264704Z",
     "start_time": "2025-11-13T14:24:34.802338Z"
    }
   },
   "source": [
    "import random\n",
    "import warnings\n",
    "\n",
    "import gunzip\n",
    "import wget\n",
    "\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "warnings.simplefilter(action='ignore', category=Warning)\n",
    "\n",
    "import Bio\n",
    "import BioSQL\n",
    "from Bio import SeqIO\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import csv\n",
    "#parse file, clean data right after\n",
    "GENOME_FILE = \"C:/Users/pixil/PycharmProjects/SimpleTorchProject\\TorchDun/data/nih_human_gproj/Homo_sapiens.GRCh38.dna.primary_assembly.fa\"\n",
    "#Final and constant, unless we change it before running ofc\n",
    "#FINAL\n",
    "FEATURE_FILE = \"C:/Users/pixil/PycharmProjects/SimpleTorchProject/TorchDun/data/nih_human_gproj/GCF_000001405.40_GRCh38.p14_feature_table.txt\"\n",
    "SEQUENCE_LENGTH = 1000  # Change sequence length\n",
    "NUM_SAMPLES = 10000     # Change number of sequences\n",
    "OUTPUT_FILE = \"data/sequences.csv\"\n",
    "\n",
    "df_feature = pd.read_csv(\n",
    "    FEATURE_FILE,\n",
    "    sep='\\t',\n",
    "    comment='#',\n",
    "    header=None,\n",
    "    usecols=[0, 1, 5, 7, 8, 9, 13, 14, 15, 19],\n",
    "    #NCIB database named\n",
    "    names=['feature', 'class', 'chromosome', 'start', 'end', \"strand\", \"description\", \"symbol\", \"GeneID\", \"attributes\"]\n",
    ")\n",
    "#our features\n",
    "print(df_feature.columns.tolist())\n",
    "def extract_sequences(GENOME_FILE, SEQUENCE_LENGTH, NUM_SAMPLES):\n",
    "    seq = []\n",
    "    for record in SeqIO.parse(GENOME_FILE, \"fasta\"):\n",
    "        chrom_name = record.id\n",
    "        chrom_seq = str(record.seq).upper() #string of sequence (letters)\n",
    "        chrom_len = len(chrom_seq) #gets our length of sequence\n",
    "        #gets all genes on the chromosome\n",
    "        chrom_genes = df_feature[(df_feature['chromosome'] == chrom_name) & (df_feature['feature'] == 'gene')]\n",
    "        print(record.id)\n",
    "        print(f\"Processing chromosome {chrom_name}: {chrom_len:,} bp\")\n",
    "        for i in range(NUM_SAMPLES):\n",
    "            if len(seq) >= NUM_SAMPLES:\n",
    "                break\n",
    "            start = random.randint(0, chrom_len - SEQUENCE_LENGTH)\n",
    "            chunk = chrom_seq[start:start + SEQUENCE_LENGTH]\n",
    "\n",
    "            #we have to check if gene is in sequence!\n",
    "            gene_snip = ((chrom_genes['start'] < start + SEQUENCE_LENGTH) & (chrom_genes['end'] > start ))\n",
    "            in_gene = gene_snip.any()\n",
    "            if in_gene:\n",
    "                gene_in_seq = chrom_genes[gene_snip]\n",
    "                gene_name = ','.join(gene_in_seq['symbol'].dropna().tolist())\n",
    "            else:\n",
    "                gene_name = ''\n",
    "            #gets bases count\n",
    "            g_count = chunk.count('G')\n",
    "            c_count = chunk.count('C')\n",
    "            a_count = chunk.count('A')\n",
    "            t_count = chunk.count('T')\n",
    "\n",
    "            #we have refseq and genseq categories as well i believe\n",
    "\n",
    "            #percent in count\n",
    "            gc_count = ((g_count + c_count) / SEQUENCE_LENGTH) * 100\n",
    "            at_count = ((a_count + t_count) / SEQUENCE_LENGTH) * 100\n",
    "            seq.append({\n",
    "                \"sequence\": chunk,\n",
    "                \"chromosome\": chrom_name,\n",
    "                \"start\": start,\n",
    "                \"end\": start + SEQUENCE_LENGTH,\n",
    "                \"gc_count\": gc_count,\n",
    "                \"at_count\": at_count,\n",
    "                \"in_gene\": in_gene,\n",
    "                \"gene_name\": gene_name\n",
    "            })\n",
    "        if len(seq) >= NUM_SAMPLES:\n",
    "            break\n",
    "\n",
    "    return seq\n",
    "sequences = extract_sequences(GENOME_FILE, SEQUENCE_LENGTH, NUM_SAMPLES)\n",
    "\n",
    "df = pd.DataFrame(sequences)\n",
    "#sequences\n",
    "print(df)\n",
    "#data now must be cleaned and put in a dataframe, set up for training+analysis\n",
    "\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['feature', 'class', 'chromosome', 'start', 'end', 'strand', 'description', 'symbol', 'GeneID', 'attributes']\n",
      "1\n",
      "Processing chromosome 1: 248,956,422 bp\n",
      "                                               sequence chromosome      start  \\\n",
      "0     AGAGCTTCAGGGAAGGAGAACCCCAGAGATCTGTAGAGGGTATTCC...          1  221576856   \n",
      "1     TAAAATTAAAAGTATAGAAATAGAAACAAAGCAATAACATTCAAGT...          1  157370272   \n",
      "2     TATTTTTTGAAACGTTGGGCAATGTCTGACTTTAGCATCTGTGTTA...          1  184341761   \n",
      "3     TTTTTGTGTATCCAATTATAGACCCTAATATCTAATTGCAGATTCT...          1  232467920   \n",
      "4     CCCCTCCCCACAGCCCCTGGCAACCACAAATCTACTTTCTGTCTCT...          1  109164203   \n",
      "...                                                 ...        ...        ...   \n",
      "9995  ATTTATAATTCATTGAGGTAAGCACACAGGGACGAAGGAGTCAGCC...          1  236687338   \n",
      "9996  GTTAATTATTACCAATAGAAGCTTATTTATGCTTGTAACTGAGTTA...          1  243365017   \n",
      "9997  ACGCCATTCTCCTGCTTCAGCCTCCCGAGTAGCTGGGACTACAGGT...          1   10372728   \n",
      "9998  GGTCTGCGATCTGAGGTCTTGCTTTAGCAAGCTCTCTTATTTAAGT...          1  236877201   \n",
      "9999  ATCTTCAATCACTGATATCCTTTCTTCCACTTGATCAAATCGGCTA...          1   37379789   \n",
      "\n",
      "            end  gc_count  at_count  in_gene gene_name  \n",
      "0     221577856      37.7      62.3    False            \n",
      "1     157371272      40.8      59.2    False            \n",
      "2     184342761      37.5      62.5    False            \n",
      "3     232468920      36.1      63.9    False            \n",
      "4     109165203      53.9      46.1    False            \n",
      "...         ...       ...       ...      ...       ...  \n",
      "9995  236688338      40.9      59.1    False            \n",
      "9996  243366017      32.1      67.9    False            \n",
      "9997   10373728      48.6      51.4    False            \n",
      "9998  236878201      35.9      64.1    False            \n",
      "9999   37380789      47.8      52.2    False            \n",
      "\n",
      "[10000 rows x 8 columns]\n"
     ]
    }
   ],
   "execution_count": 45
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-12T21:22:25.462190300Z",
     "start_time": "2025-11-12T21:19:32.529771Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "70050aeb8f9063bc",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-13T00:51:20.063362Z",
     "start_time": "2025-11-13T00:51:20.056109Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "ea697cf2079a9f4f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "47bae9f0e57b8839"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
